src/proxy-ui-api/src/test/java/org/niis/xroad/restapi/openapi/TokenCertificatesApiControllerTest.java

```diff
@@ -0,0 +1,242 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.niis.xroad.restapi.openapi;
+
+import ee.ria.xroad.common.CodedException;
+
+import lombok.extern.slf4j.Slf4j;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.niis.xroad.restapi.exceptions.ErrorDeviation;
+import org.niis.xroad.restapi.facade.GlobalConfFacade;
+import org.niis.xroad.restapi.facade.SignerProxyFacade;
+import org.niis.xroad.restapi.openapi.model.CertificateDetails;
+import org.niis.xroad.restapi.openapi.model.KeyUsage;
+import org.niis.xroad.restapi.service.CertificateAlreadyExistsException;
+import org.niis.xroad.restapi.service.ClientNotFoundException;
+import org.niis.xroad.restapi.service.KeyNotFoundException;
+import org.niis.xroad.restapi.service.TokenCertificateService;
+import org.niis.xroad.restapi.util.CertificateTestUtils;
+import org.niis.xroad.restapi.util.TestUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.core.io.Resource;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.context.junit4.SpringRunner;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.security.cert.X509Certificate;
+import java.time.OffsetDateTime;
+import java.util.ArrayList;
+import java.util.Collections;
+
+import static ee.ria.xroad.common.ErrorCodes.SIGNER_X;
+import static ee.ria.xroad.common.ErrorCodes.X_CERT_EXISTS;
+import static ee.ria.xroad.common.ErrorCodes.X_CSR_NOT_FOUND;
+import static ee.ria.xroad.common.ErrorCodes.X_INCORRECT_CERTIFICATE;
+import static ee.ria.xroad.common.ErrorCodes.X_KEY_NOT_FOUND;
+import static ee.ria.xroad.common.ErrorCodes.X_WRONG_CERT_USAGE;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.doThrow;
+
+/**
+ * test certificates api
+ */
+@RunWith(SpringRunner.class)
+@SpringBootTest
+@AutoConfigureTestDatabase
+@Transactional
+@Slf4j
+public class TokenCertificatesApiControllerTest {
+
+    @MockBean
```

Comment:
So this is now kind of an _integration test_ since it uses real TokenCertificateService and CertificateDetailsConverter. It's not wrong but
- we have sometimes used "xxIntegrationTest" naming for those, ServicesApiControllerIntegrationTest (pretty certain it's not consistent though)
- sometimes this kind of approach ends up biting you (me). I did something similar yesterday and ended up lost for a while, since I did not realize my controller used real services / real global configuration under the hood
- it's often said that you should isolate the class under test as much as possible, so that when test fails you know the problem is in the tested class and not the dependencies.
- I don't think be have to pick either / or about integration tests, but would it make things clearer if integration tests and "proper unit tests" would be clearly separated? What do you think about the naming thing (*IntegrationTest for tests that test a group of components instead of one class)....?

Reply:
Yeah we should probably be more consistent with the naming of the tests. I think it is good enough if we call controller tests that use the real injected services/converters with the suffix `...IntergrationTest`.